### PR TITLE
gh-91576: Speed up iteration of strings

### LINKED ISSUE #91576: Speed up iteration of strings
PR speeds up iteration with:
- reducing branches in the iternext function.
- Add a special ascii iterator for strings which covers the common case.

Benchmark results are attached in the PR.

### PR DESCRIPTION
**Pull request by kumaraditya303**
Benchmark Script:
```py
from pyperf import Runner, perf_counter

def bench_str(loops, src, length):
    src = src * length
    t0 = perf_counter()
    for _ in range(loops):
        for i in src:
            pass
    return perf_counter() - t0

runner = Runner()
for src in ['a', 'é']:
    for n in [1_000, 10_000]:
        runner.bench_time_func(f"str {src} {n}", bench_str, src, n)
```
Results:
```console 
str a 1000: Mean +- std dev: [base] 9.72 us +- 0.62 us -> [patch] 6.54 us +- 1.05 us: 1.49x faster
str a 10000: Mean +- std dev: [base] 95.8 us +- 5.1 us -> [patch] 63.0 us +- 4.1 us: 1.52x faster
str é 1000: Mean +- std dev: [base] 10.3 us +- 1.6 us -> [patch] 8.60 us +- 0.56 us: 1.19x faster
str é 10000: Mean +- std dev: [base] 103 us +- 9 us -> [patch] 86.1 us +- 4.7 us: 1.19x faster

Geometric mean: 1.34x faster
```

Closes https://github.com/python/cpython/issues/91576

### CODE REVIEW

File: Objects/object.c [hunk 1]
```diff
@@ -1912,6 +1914,7 @@ static PyTypeObject* static_types[] = {
     &PyTraceBack_Type,
     &PyTupleIter_Type,
     &PyTuple_Type,
+    &PyUnicodeASCIIIter_Type
```
Comment (arhadthedev) [inline Objects/object.c]:
```suggestion
    &PyUnicodeASCIIIter_Type,
```

**Comment by JelleZijlstra** [comment]
Happy to help review this, let me know when you're ready

**Comment by kumaraditya303** [comment]
> Happy to help review this, let me know when you're ready

@JelleZijlstra Finished.

**Review by gvanrossum — COMMENTED** [review]
Why not use the specialized iteratie for all Latin-1 strings?

**Comment by kumaraditya303** [comment]
> Why not use the specialized iteratie for all Latin-1 strings?

That would add one more branch instruction and I was trying to avoid it and LATIN1 is rare compared to ASCII.

File: Objects/unicodeobject.c [hunk 2]
```diff
@@ -15819,7 +15850,11 @@ unicode_iter(PyObject *seq)
     }
     if (PyUnicode_READY(seq) == -1)
         return NULL;
-    it = PyObject_GC_New(unicodeiterobject, &PyUnicodeIter_Type);
+    if (PyUnicode_IS_COMPACT_ASCII(seq)) {
+        it = PyObject_GC_New(unicodeiterobject, &_PyUnicodeASCIIIter_Type);
+    } else {
```
Comment (erlend-aasland) [inline Objects/unicodeobject.c]:
```suggestion
    }
    else {
```

File: Objects/unicodeobject.c [hunk 3]
```diff
@@ -15720,6 +15718,26 @@ unicodeiter_next(unicodeiterobject *it)
     return NULL;
 }
 
+static PyObject *
+unicode_ascii_iter_next(unicodeiterobject *it)
+{
+    assert(it != NULL);
+    PyObject *seq = it->it_seq;
+    if (seq == NULL)
+        return NULL;
```
Comment (erlend-aasland) [inline Objects/unicodeobject.c]:
```suggestion
    if (seq == NULL) {
        return NULL;
    }
```

File: Objects/unicodeobject.c [hunk 4]
```diff
@@ -15708,10 +15708,8 @@ unicodeiter_next(unicodeiterobject *it)
     if (it->it_index < PyUnicode_GET_LENGTH(seq)) {
         int kind = PyUnicode_KIND(seq);
         const void *data = PyUnicode_DATA(seq);
-        Py_UCS4 chr = PyUnicode_READ(kind, data, it->it_index);
-        item = PyUnicode_FromOrdinal(chr);
-        if (item != NULL)
-            ++it->it_index;
+        Py_UCS4 chr = PyUnicode_READ(kind, data, it->it_index++);
```
Comment (erlend-aasland) [inline Objects/unicodeobject.c]: Note that the `PyUnicode_READ` macro reuse its arguments. I'm not sure if it is possible construct strings that misuse this misfeature, though.

File: Objects/unicodeobject.c [hunk 5]
```diff
@@ -15708,10 +15708,8 @@ unicodeiter_next(unicodeiterobject *it)
...
+        item = unicode_char(chr);
         return item;
```
Comment (gvanrossum) [inline Objects/unicodeobject.c]:
```suggestion
        return unicode_char(chr);
```
Then I think you can remove the declaration of `item`.

File: Objects/unicodeobject.c (hunk 4 above)
Comment (JelleZijlstra) [reply Objects/unicodeobject.c]: That makes the `it->it_index++` call a bit suspect. Maybe we should move it out of the macro.

**Review by gvanrossum — COMMENTED** [review]
You should just be able to test
```
(PyUnicode_KIND((unicode)) == PyUnicode_1BYTE_KIND
```
to decide which iterator to create, right? Or can `kind` be changed (once the object is "ready")?

**Comment by gvanrossum** [comment]
> That would add one more branch instruction and I was trying to avoid it and LATIN1 is rare compared to ASCII.

Given that this is a fixed cost (once per iterator construction) I think the extra branch won't be noticeable. Latin-1 may be rare compared to ASCII but it's still got some common characters and it would be essentially free.

**Comment by kumaraditya303** [comment]
> Given that this is a fixed cost (once per iterator construction) I think the extra branch won't be noticeable. Latin-1 may be rare compared to ASCII but it's still got some common characters and it would be essentially free.

No, the cost is a branch instruction on each iteration as ascii and latin1 uses different structures.

**Comment by gvanrossum** [comment]
> No, the cost is a branch instruction on each iteration as ascii and latin1 uses different structures.

Hm, couldn't you just store a pointer to the array of bytes (and another to the end) rather than an index? Or is it possible that the bytes move around somehow?

**Comment by kumaraditya303** [comment]
See LATIN1 macro in unicodeobject.c.

**Comment by kumaraditya303** [comment]
It requires a check if ch is less than 128 then it uses a different array to index depending on the  comparison.

**Comment by sweeneyde** [comment]
How does this affect performance when ascii and non-ascii are mixed together in the same string?

**Comment by gvanrossum** [comment]
> It requires a check if ch is less than 128 then it uses a different array to index depending on the comparison.

Oh, I see. That's a bit unfortunate but I see your point and I guess ASCII strings are somewhat special anyways.

> How does this affect performance when ascii and non-ascii are mixed together in the same string?

In that case the representation of the whole string will not use the "compact ASCII" format and we'll be using the regular (slow) iterator.

@kumaraditya303 Please address the other review comments.

File: Objects/unicodeobject.c (hunk 5 above)
Comment (kumaraditya303) [reply Objects/unicodeobject.c]: Fixed

File: Objects/unicodeobject.c (hunk 4 above)
Comment (kumaraditya303) [reply Objects/unicodeobject.c]: See the expanded macro, the index cannot be reused.

```c
((Py_UCS4) ((kind) == PyUnicode_1BYTE_KIND ? ((const Py_UCS1 *)(data))[(it->it_index++)] : ((kind) == PyUnicode_2BYTE_KIND ? ((const Py_UCS2 *)(data))[(it->it_index++)] : ((const Py_UCS4 *)(data))[(it->it_index++)] ) ))
```

**Review by JelleZijlstra — COMMENTED** [review]
This looks correct to me but I don't see a lot of tests for string iterators, so it may be worth adding a few in test_unicode.py:
- Test that iteration produces the right sequence of characters
- Test that iterators cannot be instantiated
- Test that iterators can be pickled

File: Objects/unicodeobject.c (hunk 4 above)
Comment (gvanrossum) [reply Objects/unicodeobject.c]: Even though the macro *happens* to work correctly, I would prefer not to have to rely on knowing that. I think we can trust the optimizer to do the right thing if we increment the index after the macro call.

**Review by gvanrossum — CHANGES_REQUESTED** [review]

File: Objects/unicodeobject.c (hunk 4 above)
Comment (kumaraditya303) [reply Objects/unicodeobject.c]: done

**Comment by kumaraditya303** [comment]
Added some tests and addressed comments.

File: Lib/test/test_unicode.py [hunk 6]
```diff
@@ -185,6 +185,18 @@ def test_iterators(self):
...
+        cases = [type(iter('abc')), type(iter('🚀'))]
+        for cls in cases:
+            with self.subTest(cls=cls):
+                self.assertRaises(TypeError, cls)
+
+    def test_iteration(self):
+        cases = ['abc', '🚀🚀🚀', "\u1111\u2222\u3333"]
+        for case in cases:
+            with self.subTest(string=case):
+                self.assertEqual((*case,), (*iter(case),))
```
Comment (JelleZijlstra) [inline Lib/test/test_unicode.py]: Both sides here use iteration, so this doesn't actually test that it works correctly. Use `"".join(iter(case))` on the RHS instead.

Comment (kumaraditya303) [reply Lib/test/test_unicode.py]: I was waiting for buildbots to finish before committing more changes, I'll push the fix now.

File: Objects/unicodeobject.c [hunk 7]
```diff
@@ -15709,17 +15709,37 @@ unicodeiter_next(unicodeiterobject *it)
...
+    assert(it != NULL);
+    PyObject *seq = it->it_seq;
+    if (seq == NULL) {
+        return NULL;
+    }
+    assert(_PyUnicode_CHECK(seq));
+    assert(PyUnicode_IS_COMPACT_ASCII(seq));
+    if (it->it_index < PyUnicode_GET_LENGTH(seq)) {
+        const void *data = ((void*)(_PyASCIIObject_CAST(seq) + 1));
+        Py_UCS1 chr = PyUnicode_READ(PyUnicode_1BYTE_KIND, data, it->it_index);
```
Comment (erlend-aasland) [inline Objects/unicodeobject.c]: I get a conversion warning here.

```suggestion
        Py_UCS1 chr = (Py_UCS1)PyUnicode_READ(PyUnicode_1BYTE_KIND,
                                              data, it->it_index);
```

File: Objects/unicodeobject.c [hunk 8]
```diff
@@ -15709,17 +15709,37 @@ unicodeiter_next(unicodeiterobject *it)
...
 }
 
+static PyObject *
+unicode_ascii_iter_next(unicodeiterobject *it)
+{
+    assert(it != NULL);
+    PyObject *seq = it->it_seq;
+    if (seq == NULL) {
+        return NULL;
+    }
```
Comment (erlend-aasland) [inline Objects/unicodeobject.c:15728]: Is there a code path that can trigger this branch? If not, can we make it an `assert`?

**Review by erlend-aasland — APPROVED** [review]
Looks good!

File: Objects/unicodeobject.c (hunk 7 above)
Comment (erlend-aasland) [reply Objects/unicodeobject.c]: (also, the line is too long) #nit

File: Objects/unicodeobject.c (hunk 8 above)
Comment (sweeneyde) [reply Objects/unicodeobject.c:15728]: I think this should do it:
```python
>>> s = iter('abc')
>>> list(s)
['a', 'b', 'c']
>>> list(s) # here
[]
```

File: Lib/test/test_unicode.py [hunk 9]
```diff
@@ -185,6 +185,18 @@ def test_iterators(self):
...
+        for cls in cases:
+            with self.subTest(cls=cls):
+                self.assertRaises(TypeError, cls)
+
+    def test_iteration(self):
+        cases = ['abc', '🚀🚀🚀', "\u1111\u2222\u3333"]
+        for case in cases:
+            with self.subTest(string=case):
+                self.assertEqual(case, "".join(iter(case)))
+
```
Comment (sweeneyde) [inline Lib/test/test_unicode.py:200]: Maybe add a test that already-exhausted iterators continue raising StopIteration over and over; test the `if (seq == NULL)` branch.

File: Objects/unicodeobject.c [hunk 10]
```diff
@@ -15819,7 +15852,12 @@ unicode_iter(PyObject *seq)
     }
     if (PyUnicode_READY(seq) == -1)
         return NULL;
-    it = PyObject_GC_New(unicodeiterobject, &PyUnicodeIter_Type);
+    if (PyUnicode_IS_COMPACT_ASCII(seq)) {
+        it = PyObject_GC_New(unicodeiterobject, &_PyUnicodeASCIIIter_Type);
```
Comment (sweeneyde) [inline Objects/unicodeobject.c:15857]: Is it possible for a `unicodeiterobject` to participate in a reference cycle? Could we get rid of the GC headers?

Comment (sweeneyde) [reply Objects/unicodeobject.c:15857]: Ah yes, str subclasses can have cyclic references.

File: Objects/unicodeobject.c (hunk 8 above)
Comment (erlend-aasland) [reply Objects/unicodeobject.c:15728]: In that case, we need a test for that; this branch is not exercised in the test suite.

File: Lib/test/test_unicode.py (hunk 9 above)
Comment (JelleZijlstra) [reply Lib/test/test_unicode.py:200]: Also add a test for pickling iterator objects to exercise the `unicodeiter_reduce` and `unicodeiter_setstate` functions.

Comment (kumaraditya303) [reply Lib/test/test_unicode.py:200]: done

File: Objects/unicodeobject.c (hunk 8 above)
Comment (kumaraditya303) [reply Objects/unicodeobject.c:15728]: done

File: Objects/unicodeobject.c (hunk 7 above)
Comment (kumaraditya303) [reply Objects/unicodeobject.c]: done

**Review by sweeneyde — APPROVED** [review]

**Review by gvanrossum — APPROVED** [review]
I'm happy now!

**Pull request merged**